Add aggregation and function objects to the query builder.

maltzj opened the pull request:
BigQuery has the concept of aggregation scopes, which weren't supported previously.
Also, the previous model of functions did not really jive with having if statements,
since they can have multiple parameters.  Rather than continue the old model of inferring
function calls from strings, I moved to a more OO model of declaring and applying functions.

This will require forking the library, so I'd be open to suggestions about how
to have the previous model and this one coexist.

maltzj committed:
Make it easier to develop and test the library.

Building the tests requires having pulled all the relevant dependencies from
pypi, so we may as well codify that in the Makefile.  I also added venv/ to
.gitignore so that it developers using a virtualenv don't need to think about
it.

maltzj committed:
Add support for IF formatters and aggregation over a scope.

This adds support for generating IF statements as a formatter.  This
required extending the formatter logic a bit and making it non-trivial.  To hide
that complexity, I added a generate_formatter function which will create
formatter strings for anyone consuming this library.

bkbarry commented on bigquery/query_builder.py:
```diff
@@ -84,14 +101,23 @@ def _render_select(selections):
         for options_dict in options:
             name = original_name
+            name_qualifiers = list()
+            aggregation_level = options_dict.get('aggregation_level')
```
Not thrilled about this make-a-list-and-append structure, but you're sorta forced into it with this whole options dict notion

maltzj replied:
That makes two of us.

bkbarry commented on bigquery/query_builder.py:
```diff
@@ -84,14 +101,23 @@ def _render_select(selections):
                 name = _format_select(formatter, name)
 
-            rendered_selections.append("%s %s" % (name, alias))
+            rendered_selections.append("%s %s" % (name,
```
nit: `'str {0}'.format(blah)` is preferred (% is on the road toward python deprecation)

bkbarry commented on bigquery/tests/test_query_builder.py:
```diff
@@ -3,6 +3,40 @@
+        self.assertEqual(generate_formatter('MAX'), 'MAX')
+
+    def test_single_function_with_args(self):
+        from bigquery.query_builder import generate_formatter
```
Can we just import this at the top of the file?

bkbarry commented on bigquery/tests/test_query_builder.py:
```diff
@@ -3,6 +3,40 @@
+    def test_single_function_with_args(self):
+        from bigquery.query_builder import generate_formatter
+
+        argument_list = list()
```
Do we need to store this as temporary state? We can probably do something like:

```
self.assertEqual(
  generate_formatter(
    'LEFT',
    arguments=[10, 'test']
...
```

bkbarry commented on bigquery/tests/test_query_builder.py:
```diff
@@ -3,6 +3,40 @@
+    def test_nesting_with_args(self):
+        from bigquery.query_builder import generate_formatter
+        
+        argument_list = list()
```
Cool thing about python: list construction can be declarative -- no need to make the object and mutate it. Something like:
`foo = [bang, bar]`

bkbarry commented:
I am indebted for taking so long to get to this. It looks great -- only suggestion would be looking over the new code and culling any unnecessary state that might've snuck in.

(Also, holler whenever you get to this and I'll put it to the top of my todo list)

bkbarry commented on bigquery/tests/test_query_builder.py:
```diff
@@ -52,6 +86,63 @@ def test_no_selects(self):
 
         self.assertEqual(result, 'SELECT *')
 
+    def test_aggregation_within_record(self):
```
These're great tests, which is super important given the amount of string manipulation that we're doing

bkbarry commented on bigquery/tests/test_query_builder.py:
```diff
@@ -52,6 +76,63 @@ def test_no_selects(self):
+        Ensure that aggregation within a query works when there is no alias for
+        that field.
+        """
+        from bigquery.query_builder import _render_select 
```
Why not import this once at the top-most level?

bkbarry commented:
Optional comment, shipit when you're happy :D

maltzj committed:
Move all imports to the top level in tests.

It didn't seem valuable to repeat this in every test, so I just made them
top-level imports.
